risc0/zkvm/src/groth16.rs

```diff
@@ -0,0 +1,802 @@
+    fn try_from(raw_proof: RawProof) -> Result<Self, Error> {
+        if raw_proof.pi_a.len() < 2 {
+            bail!("Malformed G1 element field");
+        }
+        let a = vec![
+            from_u256(&raw_proof.pi_a[0])?,
+            from_u256(&raw_proof.pi_a[1])?,
+        ];
+
+        if raw_proof.pi_b.len() < 2 || raw_proof.pi_b[0].len() < 2 || raw_proof.pi_b[1].len() < 2 {
+            bail!("Malformed G2 element field");
+        }
+        let b = vec![
+            vec![
+                from_u256(&raw_proof.pi_b[0][1])?,
+                from_u256(&raw_proof.pi_b[0][0])?,
+            ],
+            vec![
+                from_u256(&raw_proof.pi_b[1][1])?,
+                from_u256(&raw_proof.pi_b[1][0])?,
+            ],
+        ];
+
+        if raw_proof.pi_c.len() < 2 {
+            bail!("Malformed G1 element field");
+        }
+        let c = vec![
+            from_u256(&raw_proof.pi_c[0])?,
+            from_u256(&raw_proof.pi_c[1])?,
+        ];
+
+        Ok(Groth16Seal { a, b, c })
+    }
+}
+
+/// Groth16 instance over the BN_254 curve encoded in little endian
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Groth16 {
+    pvk: Vec<u8>,
+    proof: Vec<u8>,
+    prepared_inputs: Vec<u8>,
+}
+
+impl Groth16 {
+    /// Creates a Groth16 instance from a `Groth16Seal` and the metadata digest
+    /// of the original RISC Zero receipt
+    #[cfg(not(target_os = "zkvm"))]
+    pub fn from_seal(groth16_seal: &Groth16Seal, receipt_meta: Digest) -> Result<Self, Error> {
+        let mut pvk_bytes = Vec::new();
+        let public_key_verification = Self::pvk()?;
+        public_key_verification
+            .serialize_uncompressed(&mut pvk_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        let mut proof_bytes = Vec::new();
+        let proof = Proof::<Bn254> {
+            a: g1_from_bytes(&groth16_seal.a)?,
+            b: g2_from_bytes(&groth16_seal.b)?,
+            c: g1_from_bytes(&groth16_seal.c)?,
+        };
+        proof
+            .serialize_uncompressed(&mut proof_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        let mut prepared_inputs_bytes = Vec::new();
+        let (c1, c2) =
+            split_digest(Digest::from_hex(ALLOWED_IDS_ROOT).map_err(|err| anyhow!(err))?)?;
+        let (m1, m2) = split_digest(receipt_meta)?;
+        let public_inputs = vec![c2, c1, m2, m1];
+        let prepared_inputs =
+            ark_Groth16::<Bn254>::prepare_inputs(&public_key_verification, &public_inputs)
+                .map_err(|err| anyhow!(err))?;
+        prepared_inputs
+            .serialize_uncompressed(&mut prepared_inputs_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        Ok(Self {
+            pvk: pvk_bytes,
+            proof: proof_bytes,
+            prepared_inputs: prepared_inputs_bytes,
+        })
+    }
+
+    /// Creates a Groth16 instance from the raw material generated by Circom/SnarkJS
+    #[cfg(not(target_os = "zkvm"))]
+    pub fn from_raw(
```

Comment:
Maybe instead of "raw", this should be "circom"?
```suggestion
    pub fn from_circom(
```